File: cmd/client/server/logout.go
package server

import "net/http"

// Logout handles user logout by clearing session cookies.
func (cs *ClientServer) Logout(w http.ResponseWriter, r *http.Request) {
	cs.clearSessionCookies(w)

	// Redirect to homepage
	http.Redirect(w, r, "/", http.StatusSeeOther)
}

// clearSessionCookies clears both session cookies (for logout).
func (cs *ClientServer) clearSessionCookies(w http.ResponseWriter) {
	accessCookie := &http.Cookie{
		Name:     "access_token",
		Value:    "",
		Path:     "/",
		HttpOnly: true,
		Secure:   false,
		SameSite: http.SameSiteLaxMode,
		MaxAge:   -1,
	}

	refreshCookie := &http.Cookie{
		Name:     "refresh_token",
		Value:    "",
		Path:     "/",
		HttpOnly: true,
		Secure:   false,
		SameSite: http.SameSiteLaxMode,
		MaxAge:   -1,
	}

	http.SetCookie(w, accessCookie)
	http.SetCookie(w, refreshCookie)
}

File: frontend/static/js/pages/logout.js
/**
 * pages/logout.js
 *
 * Calls backend POST /api/v1/logout, clears client auth state,
 * re-renders the navbar in guest mode, then navigates home.
 *
 * The BFF's Logout handler did:
 *   1. POST to backend /logout (with cookies so the session is deleted server-side)
 *   2. Clear the access_token and refresh_token cookies (MaxAge: -1)
 *   3. Redirect to /
 *
 * In CSR the cookie clearing happens server-side when the backend logout
 * endpoint responds — it sets Set-Cookie headers with MaxAge:-1, which the
 * browser processes automatically because we send credentials:"include".
 */

import { api } from '../api.js';
import { navigate } from '../router.js';
import { closeNotificationStream } from '../navbar.js';

export async function renderLogoutPage(user) {
  const root = document.getElementById('app-root');
  if (!root) return;

  if (!user) {
    navigate('/login');
    return;
  }

  root.innerHTML = buildHTML('Signing out…');

  try {
    // POST /api/v1/logout — backend deletes the session and clears cookies
    await api.post('/logout');
  } catch (err) {
    console.warn('Backend logout failed, clearing client state anyway:', err.message);
  }

  //the cookies are deleted from the backend response, so we just go to the home page and the middleware will get the user and render the navbar

  closeNotificationStream();

  navigate('/');
}

function buildHTML(message, showRetry = false) {
  return /* html */ `
    <div class="main-container" style="text-align:center;padding:4rem 1rem">
      <div class="page-loading">
        <div class="page-loading-spinner"></div>
      </div>
      <p style="margin-top:1.5rem;color:var(--white-background-light);font-size:1.1rem">
        ${message}
      </p>
      ${
        showRetry
          ? `<div style="margin-top:1.5rem">
             <button id="logout-retry" class="btn-signup">Retry</button>
           </div>`
          : ''
      }
    </div>
  `;
}
